### DISCUSSION TITLE
Clarification for execution order of route `onEnter` and `onLeave`

### DISCUSSION THREAD

**Discussion by LittorWired** [TanStack/router#2681] [Q&A]
I am hoping to get some clarification on the exact order and conditions for `onEnter`, `onStay`, and `onLeave` in the `createFileRoute` methods, specifically:
1. In a route transition to a different route
2. In a route transition with the same route but different dynamic parameters

Are there any guarantees about their execution order?

Specifically, in our app, we observed that in the case of transitioning to the same route with different parameters, `onEnter` appears to be called before `onLeave`, which seems counterintuitive to the method descriptions. Is this the intended behavior?

https://tanstack.com/router/latest/docs/framework/react/api/router/RouteOptionsType#onleave-property

**Comment by schiller-manuel** [comment]
> Specifically, in our app, we observed that in the case of transitioning to the same route with different parameters, onEnter appears to be called before onLeave

 can you provide a complete minimal example to debug this?

**Reply by LittorWired** [reply]
Apologies, the execution order for `onEnter`, `onStay`, and `onLeave` is working as expected. Must've not had enough rest recently 🤦 .

The question is regarding the execution order with the `loader` in the context of `onEnter` and `onLeave` . I will close this issue and reopen a new discussion with a minimal example and the question corrected. Thank you
